core/tests/test_utils.py

```diff
@@ -1143,12 +1143,17 @@ def swap(self, obj, attr, newvalue):
         the generator will immediately raise StopIteration, and contextlib will
         raise a RuntimeError.
         """
-        original = getattr(obj, attr)
+        had_attribute = hasattr(obj, attr)
```

Comment:
The logic change here is odd given the name. If you're adding an attribute then you're not actually "swapping" an existing attribute out.

Can you avoid making this change? Where is it needed?

Reply:
This is needed because some virtual environment tools set `sys.real_path` to signal that we are in a virtual environment (i.e. if `sys.real_path` exists, we are in a virtual environment). To test this check, we need to temporarily set `sys.real_path` and then after the test, return it to its original state (either set it back to its original value or, if it wasn't originally present, delete it)

Reply:
Hm, interesting. In that case, instead of modifying self.swap(), would it be possible to add another utility method that has the semantics of "set sys.real_path context", and leave the original self.swap() unchanged? I think what is being done in this case is not a swap and I don't think we should be extending the function's contract to the case where the property doesn't exist (in general) since this is very much a special case.

Reply:
done